adr/0016-hooks.md

```diff
@@ -0,0 +1,113 @@
+ - NOTE: The `package` lifecycle might be changed to a `run-once` lifecycle. This would benefit packages that don't have kube context information when the deployment starts.
+
+Zarf hooks will have two forms of execution via `Internal` and `External` hooks.
+### Internal Hooks
+Internal hooks will be hooks that are built into the Zarf CLI and run internal code when executed. The logic for these hooks would be built into the Zarf CLI and would be updated with new releases of the CLI.
+
+### External Hooks
+There are a few approaches for external hooks.
+1. Have the hook metadata reference an OCI image that is downloaded and run.
+ - The hook metadata can reference the shasum of the image to ensure the image is not tampered with.
+ - We can pass metadata from the secret to the image.
+2. Have the hook metadata reference an image/endpoint that we call via a gRPC call.
+ - This would require a lot of consideration to security since we will be executing code from an external source.
+3. Have the hook metadata contain a script or list of shell commands that can get run.
+ - This would be the simplest solution but would require the most work from the hook creator. This also has the most potential security issues.
+
+
+
+**PROS**
+ - Implementing Hooks internally means we don't have to deal with any bootstrapping issues.
+ - Internally managed hooks can leverage Zarf internal code.
+
+**CONS**
+ - Since 'Internal' hooks are built into the CLI, the only way to get updates for the hook is to update the CLI.
+ - External hooks will have a few security concerns that we will have to work through.
+
+
+
+### Pepr
+Pepr is a K8s controller that enables Kubernetes mutations. We are (or will be) considering using Pepr to replace the `Zarf Agent`. Pepr is capable to accomplishing most of what Zarf wants to do with the concept of Hooks. Zarf hook configuration could be saved as secrets that Zarf will be able to use
+
+**PROS**
+ - Pepr as a solution would be more ?flexible?? than internal Zarf implementation of Hooks.
+ - Using Pepr would reduce the complexity of Zarf's codebase.
+ - It will be easier to secure third party hooks when Pepr is the one running them.
+ - Lifecycle management would be a lot easier with a webhook solution like Pepr.
+
+**CONS**
+ - Pepr is a new project that hasn't been stress tested in production yet (but neither has Hooks).
+ - The Pepr image needs to be pushed to an image registry before it is deployed. This will require a new bootstrapping solution to solve the ECR problem we identified above.
+
+
+## Consequences
+
+- External hooks will likely not be implemented in the first pass of this feature. Handling external hooks will be an interesting challenge as we'll have to download the hook image and run / execute it. Security will be something we consider heavily when implementing this feature.
```

Comment:
I think it might be worth exploring both - if we don't want to deal with Pepr right away we could have an uglier expressjs server instead (since that is what Pepr would ultimately do under the hood) - that way we can get feedback / evaluate both approaches with a bit more data.

Reply:
I agree, that it'd be worth exploring both, but I feel like we might as well try with Pepr since that seems like the most flushed out solution that we would actually want to use.

Was there something you wanted me to add to the ADR here?